# Patch for built-in PBM image loader, reported at
#
#    https://debbugs.gnu.org/cgi/bugreport.cgi?bug=81344
#
# The following patch was ported back to Emacs 30.2 from the original
# found in that report.
# -----
# From: Paul Eggert <eggert@cs.ucla.edu>
# Date: Mon, 6 Jul 2026 13:17:17 -0700
# Subject: [PATCH] Fix undefined behavior in pbm_load
#
# int*int problem reported by Tristan Madani in:
# https://bugs.gnu.org/81344
# * src/image.c (pbm_load): Avoid undefined behavior when
# multiplying ints, or when adding pointer to int.
--- emacs-30.2.orig/src/image.c	2025-01-04 04:05:30.000000000 -0700
+++ emacs-30.2/src/image.c	2026-08-26 09:52:25.568317491 -0600
@@ -7701,19 +7701,22 @@
     }
   else
     {
-      int expected_size = height * width;
       bool two_byte = 255 < max_color_idx;
-      if (two_byte)
-	expected_size *= 2;
-      if (type == PBM_COLOR)
-	expected_size *= 3;
 
-      if (raw_p && p + expected_size > end)
+      if (raw_p)
 	{
-	  image_destroy_x_image (ximg);
-	  image_clear_image (f, img);
-	  image_error ("Invalid image size in image `%s'", img->spec);
-	  goto error;
+	  ptrdiff_t expected_size;
+	  bool bad = ckd_mul (&expected_size, height, width);
+	  bad |= ckd_mul (&expected_size, expected_size,
+			  (two_byte ? 2 : 1) * (type == PBM_COLOR ? 3 : 1));
+	  bad |= end - p < expected_size;
+	  if (bad)
+	    {
+	      image_destroy_x_image (ximg);
+	      image_clear_image (f, img);
+	      image_error ("Invalid image size in image `%s'", img->spec);
+	      goto error;
+	    }
 	}
 
       for (y = 0; y < height; ++y)
